### PR TITLE
Virtual height modularization

### PR DESCRIPTION
**Pull request by mts299**
# Scope 

This is origin #187 before it was merged in by accident, I think. 

@carleyjmartin you can make your comments here that you had before.

### CODE REVIEW

**Comment by carleyjmartin** [comment]
@francistee26 We thing the issues I've been having originated in develop branch (which Marina has just fixed!), but there's a few merge conflicts with those changes, if you want to make sure it's still doing what you want it to do with the new changes that fix the errors, I'll test again asap and we can get this into the next release :D

**Comment by francistee26** [comment]
> @francistee26 We think the issues I've been having originated in develop branch (which Marina has just fixed!), but there's a few merge conflicts with those changes, if you want to make sure it's still doing what you want it to do with the new changes that fix the errors, I'll test again asap and we can get this into the next release :D

@carleyjmartin that sounds good.

**Comment by francistee26** [comment]
> > @francistee26 We think the issues I've been having originated in develop branch (which Marina has just fixed!), but there's a few merge conflicts with those changes, if you want to make sure it's still doing what you want it to do with the new changes that fix the errors, I'll test again asap and we can get this into the next release :D
> 
> @carleyjmartin that sounds good.

@carleyjmartin I am confused as to which conflict of rtp.py you are talking about? Do you mean there is a conflict with rtp.py in the develop branch?

**Comment by carleyjmartin** [comment]
@francistee26 Yeah essentially, I just mean the merge conflict stopping us from merging this branch to the current develop branch. If you're in browser on github you can resolve them just below instead of on command line which is a little easier. 

The fix for the errors that is in the current develop branch is in the same lines as some changes you made, so git doesn't know how to merge them together. (I don't know how much you do or don't know about git and using github so please tell me if you either need any help or I'm just telling you what you already know!)

We essentially just want to get that bug fix into this branch, so that we can test this branch without the errors and then merge these changes it into develop. But to do that we need to resolve the merge conflicts by telling git what the code should be.

**Comment by francistee26** [comment]
> @francistee26 Yeah essentially, I just mean the merge conflict stopping us from merging this branch to the current develop branch. If you're in browser on github you can resolve them just below instead of on command line which is a little easier.
> 
> The fix for the errors that is in the current develop branch is in the same lines as some changes you made, so git doesn't know how to merge them together. (I don't know how much you do or don't know about git and using github so please tell me if you either need any help or I'm just telling you what you already know!)
> 
> We essentially just want to get that bug fix into this branch, so that we can test this branch without the errors and then merge these changes it into develop. But to do that we need to resolve the merge conflicts by telling git what the code should be.

Hi @carleyjmartin thanks I just fixed the merge conflict.

**Comment by carleyjmartin** [comment]
Great, thank you! I will test this afternoon!

**Review by carleyjmartin — CHANGES_REQUESTED** [review]
So when reviewing, I kinda looked at the error and tried to figure out what was going on. So you might see halfway through a comment I've figured out what is going on, I've left it as a stream of me figuring it out, so you can see where I've been looking and what errors I've been getting and why.

In addition to the issues with plot_fan, I'm also having some problem with plot_fov as this function is designed to be able to be used without data, and that the default is now coords.AACGM for some of the methods, it's unable to plot anything. 

Good news is the fix from develop worked and theres no further issue in any of the range time plots.

**Comment by mts299** [comment]
@francistee26 I am also having issues running the plot_fan code:

``` python
import pydarn
from datetime import datetime
import matplotlib.pyplot as plt 

rkn_file = 'data/20180101.0000.01.rkn.fitacf'
rkn_data = pydarn.SuperDARNRead().read_dmap(rkn_file)
pydarn.Fan.plot_fan(rkn_data, scan_index=3,
                    colorbar=False, line_color='blue',
                    radar_label=True)
````

```` bash
Traceback (most recent call last):
  File "plot_fan.py", line 20, in <module>
    radar_label=True)
  File "/home/marina/.local/lib/python3.6/site-packages/pydarn/plotting/fan.py", line 209, in plot_fan
    date=date, max_beams=None, read_file=fov_files)
  File "/home/marina/.local/lib/python3.6/site-packages/pydarn/utils/radar_pos.py", line 143, in radar_fov
    250, date))
  File "/usr/lib/python3.6/site-packages/aacgmv2-2.6.1-py3.6-linux-x86_64.egg/aacgmv2/wrapper.py", line 433, in get_aacgm_coord
    method_code=method_code)
  File "/usr/lib/python3.6/site-packages/aacgmv2-2.6.1-py3.6-linux-x86_64.egg/aacgmv2/wrapper.py", line 219, in convert_latlon
    if abs(in_lat) > 90.0:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Comment by francistee26** [comment]
> @francistee26 I am also having issues running the plot_fan code:
> 
> ```python
> import pydarn
> from datetime import datetime
> import matplotlib.pyplot as plt 
> 
> rkn_file = 'data/20180101.0000.01.rkn.fitacf'
> rkn_data = pydarn.SuperDARNRead().read_dmap(rkn_file)
> pydarn.Fan.plot_fan(rkn_data, scan_index=3,
>                     colorbar=False, line_color='blue',
>                     radar_label=True)
> ```
> 
>  ```shell
> Traceback (most recent call last):
>   File "plot_fan.py", line 20, in <module>
>     radar_label=True)
>   File "/home/marina/.local/lib/python3.6/site-packages/pydarn/plotting/fan.py", line 209, in plot_fan
>     date=date, max_beams=None, read_file=fov_files)
>   File "/home/marina/.local/lib/python3.6/site-packages/pydarn/utils/radar_pos.py", line 143, in radar_fov
>     250, date))
>   File "/usr/lib/python3.6/site-packages/aacgmv2-2.6.1-py3.6-linux-x86_64.egg/aacgmv2/wrapper.py", line 433, in get_aacgm_coord
>     method_code=method_code)
>   File "/usr/lib/python3.6/site-packages/aacgmv2-2.6.1-py3.6-linux-x86_64.egg/aacgmv2/wrapper.py", line 219, in convert_latlon
>     if abs(in_lat) > 90.0:
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
> ```
> ```

@mts299 I will take a look at it tomorrow and keep you updated.

File: pydarn/utils/radar_pos.py [hunk 1]
```diff
@@ -119,13 +126,14 @@ def radar_fov(stid: int, rsep: int = 45, frang: int = 180,
...
+        beam_corners_lats = np.zeros((ranges[1]+1, max_beams+1))
+        beam_corners_lons = np.zeros((ranges[1]+1, max_beams+1))
 
-        for beam in range(0, max_beams+1):
-            for gate in range(ranges[0], ranges[1]):
+        for beam in range(max_beams+1):
+            for gate in range(ranges[1]+1):
                 lat, lon = geographic_cell_positions(stid, beam, gate, rsep,
-                                                     frang, height=300)
+                                                     frang, coords = coords, reflection_height=reflection_height, height=300)
```
[the remark this answers is not in the document]
Comment (francistee26) [reply pydarn/utils/radar_pos.py:136]: @carleyjmartin and @mts299 I was able to reproduce the error you got for plot_fan. For plot_fan to work I have it set up where you need to pass in a coord type. As regards to plot_fov I did not use that method in plot_fan I instead used radar_fov.

File: pydarn/plotting/rtp.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Copyright (C) 2019 SuperDARN Canada, University of Saskwatchewan
 # Author: Marina Schmidt
 # This code is improvement based on rti.py in the DaVitpy library
-# https://github.com/vtsuperdarn/davitpy/blob/master/davitpy
+# https://github.com/vtsuperdarn/davitpy/blob/master/davitpyh
```
[the remark this answers is not in the document]
Comment (francistee26) [reply pydarn/plotting/rtp.py:4]: @carleyjmartin I did not make any changes on plot_fov. But I think in the plot_fov where the radar_fov method is called the fov_files  should be set to fov_files=fov_files for the coord Coords.AACGM if no data is used when calling the plot_fov method. I was able to make plot_fov work using plot_fov(20, datetime(2015, 3, 8, 15, 0), radar_label = True, fov_files=True) after I set the fov_files=fov_files in the radar_fov method call.

File: pydarn/utils/radar_pos.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carleyjmartin) [reply pydarn/utils/radar_pos.py:136]: Okay so for plot_fan, we will need to use default values in the code that doesn't break if the user doesn't give a specific set of inputs. The default coord at the moment is AACGM, which breaks the aacgm2 function it calls. Is this code not designed to be used with aacgm?

Further on from this, when specifying the coords using slant range or ground scatter mapped range, this results in a cartopy error. Is this PR supposed to introduce cartopy to the package? I thought that was going to be a different PR, and is there a back up for when the user doesn't have/doesn't want cartopy installed?

File: pydarn/plotting/rtp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carleyjmartin) [reply pydarn/plotting/rtp.py:4]: radar_pos and radar_fov should ideally be able to return a fov made using any of the coordinate systems, including aacgm, without needing to use the fov_files=True option right?

**Pull request closed without merging**